File: zad-web/src/utils/answerParser.ts
export enum AnswerItemType {
  paragraph,
  listItem,
  quote
}

export interface AnswerItem {
  type: AnswerItemType;
  text: string;
}

export interface AnswerSection {
  title: string;
  items: AnswerItem[];
}

export interface ParsedAnswer {
  intro?: string;
  sections: AnswerSection[];
}

export class AnswerParser {
  static parse(raw: string): ParsedAnswer {
    const lines = raw.split('\n');

    let intro: string | undefined = undefined;
    const sections: AnswerSection[] = [];
    let current: AnswerSection | null = null;

    let introBuffer = '';
    let foundFirstSection = false;

    for (let i = 0; i < lines.length; i++) {
      const rawLine = lines[i];
      const cleanLine = rawLine.trim();

      // Check if this line is a section header
      let isHeader = false;
      let headerTitle = '';

      if (cleanLine.startsWith('###')) {
        isHeader = true;
        headerTitle = cleanLine.replace(/^#+\s*/, '').trim();
      } else if (cleanLine.startsWith('**') && 
                 (cleanLine.endsWith('**') || cleanLine.endsWith(':**') || cleanLine.endsWith(':'))) {
        isHeader = true;
        headerTitle = cleanLine.replace(/\*\*/g, '').replace(/:$/, '').trim();
      }

      if (isHeader) {
        foundFirstSection = true;
        if (current !== null) {
          sections.push(current);
        }
        current = { title: headerTitle, items: [] };
      } else if (!foundFirstSection) {
        if (cleanLine === '') {
          introBuffer += '\n\n';
        } else {
          if (introBuffer.length > 0 && !introBuffer.endsWith('\n\n')) {
            introBuffer += '\n';
          }
          introBuffer += rawLine;
        }
      } else {
        if (cleanLine === '') {
          current?.items.push({ type: AnswerItemType.paragraph, text: '' });
          continue;
        }

        const bulletMatch = cleanLine.match(/^[\*\-\+]\s+(.+)$/);
        const numberedMatch = cleanLine.match(/^\d+\.\s*(.+)$/);
        const quoteMatch = cleanLine.match(/^>\s*(.+)$/);

        if (bulletMatch) {
          current?.items.push({ type: AnswerItemType.listItem, text: bulletMatch[1].trim() });
        } else if (numberedMatch) {
          current?.items.push({ type: AnswerItemType.listItem, text: cleanLine });
        } else if (quoteMatch) {
          current?.items.push({ type: AnswerItemType.quote, text: quoteMatch[1].trim() });
        } else {
          current?.items.push({ type: AnswerItemType.paragraph, text: cleanLine });
        }
      }
    }

    if (current !== null) {
      sections.push(current);
    }

    const introText = introBuffer.trim();
    if (introText.length > 0) {
      intro = introText;
    }

    return { intro, sections };
  }
}

File: AI-Powered-Islamic-Educational-and-QA-System-using-RAG-Mobile-App-Flutter/lib/features/chatbot/presentation/utils/answer_parser.dart
import 'package:flutter/material.dart';

/// Parses the raw `answer` string from the API into structured sections.
class AnswerParser {
  static ParsedAnswer parse(String raw) {
    final lines = raw.split('\n');

    String? intro;
    final sections = <AnswerSection>[];
    AnswerSection? current;

    final introBuffer = StringBuffer();
    bool foundFirstSection = false;

    for (int i = 0; i < lines.length; i++) {
      final rawLine = lines[i];
      final cleanLine = rawLine.trim();

      // Check if this line is a section header
      bool isHeader = false;
      String headerTitle = '';

      if (cleanLine.startsWith('###')) {
        isHeader = true;
        headerTitle = cleanLine.replaceFirst(RegExp(r'^#+\s*'), '').trim();
      } else if (cleanLine.startsWith('**') &&
          (cleanLine.endsWith('**') ||
              cleanLine.endsWith(':**') ||
              cleanLine.endsWith(':'))) {
        isHeader = true;
        headerTitle = cleanLine.replaceAll('**', '').trim();
      }

      if (isHeader) {
        foundFirstSection = true;
        if (current != null) sections.add(current);
        current = AnswerSection(title: headerTitle, items: []);
      } else if (!foundFirstSection) {
        if (cleanLine.isEmpty) {
          introBuffer.write('\n\n');
        } else {
          if (introBuffer.isNotEmpty &&
              !introBuffer.toString().endsWith('\n\n')) {
            introBuffer.write('\n');
          }
          introBuffer.write(rawLine);
        }
      } else {
        if (cleanLine.isEmpty) {
          current?.items.add(const AnswerItem(
            type: AnswerItemType.paragraph,
            text: '',
          ));
          continue;
        }

        final bulletMatch = RegExp(r'^[\*\-\+]\s+(.+)$').firstMatch(cleanLine);
        final numberedMatch = RegExp(r'^\d+\.\s*(.+)$').firstMatch(cleanLine);
        final quoteMatch = RegExp(r'^>\s*(.+)$').firstMatch(cleanLine);

        if (bulletMatch != null) {
          current?.items.add(AnswerItem(
            type: AnswerItemType.listItem,
            text: bulletMatch.group(1)!.trim(),
          ));
        } else if (numberedMatch != null) {
          current?.items.add(AnswerItem(
            type: AnswerItemType.listItem,
            text: cleanLine, // Keep the number prefix
          ));
        } else if (quoteMatch != null) {
          current?.items.add(AnswerItem(
            type: AnswerItemType.quote,
            text: quoteMatch.group(1)!.trim(),
          ));
        } else {
          current?.items.add(AnswerItem(
            type: AnswerItemType.paragraph,
            text: cleanLine,
          ));
        }
      }
    }

    if (current != null) sections.add(current);

    final introText = introBuffer.toString().trim();
    if (introText.isNotEmpty) intro = introText;

    return ParsedAnswer(intro: intro, sections: sections);
  }

  /// Extracts citation keys referenced in a text, e.g. `[cit_2]` → `cit_2`,
  /// or `[2]` → key lookup by index.
  static List<String> extractCitationRefs(String text) {
    final refs = <String>[];
    final pattern = RegExp(r'\[(cit_\d+|\d+)\]');
    for (final m in pattern.allMatches(text)) {
      refs.add(m.group(1)!);
    }
    return refs;
  }

  /// Strips citation markers from display text (handles multi-cites like [3, 5] as well).
  static String stripCitationRefs(String text) =>
      text.replaceAll(RegExp(r'\s*\[[^\]]+\]'), '').trim();

  static List<TextSpan> parseRichText(String text, TextStyle baseStyle) {
    final List<TextSpan> spans = [];
    final regex = RegExp(r'\*\*(.*?)\*\*|\[(.*?)\]');
    int start = 0;

    for (final match in regex.allMatches(text)) {
      if (match.start > start) {
        spans.add(TextSpan(
          text: text.substring(start, match.start),
          style: baseStyle,
        ));
      }
      
      final boldText = match.group(1);
      final citeText = match.group(2);

      if (boldText != null) {
        spans.add(TextSpan(
          text: boldText,
          style: baseStyle.copyWith(fontWeight: FontWeight.bold),
        ));
      } else if (citeText != null) {
        if (RegExp(r'^(cit_\d+|\d+)(,\s*(cit_\d+|\d+))*$').hasMatch(citeText)) {
          final displayNum = citeText.replaceAll('cit_', '');
          spans.add(TextSpan(
            text: ' [$displayNum] ',
            style: baseStyle.copyWith(
              color: const Color(0xFFBA68C8),
              fontWeight: FontWeight.bold,
            ),
          ));
        } else {
           spans.add(TextSpan(
             text: '[$citeText]',
             style: baseStyle,
           ));
        }
      }
      start = match.end;
    }

    if (start < text.length) {
      spans.add(TextSpan(
        text: text.substring(start),
        style: baseStyle,
      ));
    }

    return spans;
  }
}

class ParsedAnswer {
  final String? intro;
  final List<AnswerSection> sections;

  const ParsedAnswer({this.intro, required this.sections});
}

class AnswerSection {
  final String title;
  final List<AnswerItem> items;

  AnswerSection({required this.title, required this.items});
}

class AnswerItem {
  final AnswerItemType type;
  final String text;

  const AnswerItem({required this.type, required this.text});
}

enum AnswerItemType { paragraph, listItem, quote }
